### ISSUE TITLE
Notebook testing automation

### ISSUE THREAD

**Issue by jklaise** [SeldonIO/alibi#476]
Currently notebooks are only infrequently tested by manual execution. To ensure that they are up-to-date we should included this in our standard testing process.

By "testing" we mean:

- Can the notebook be executed without errors?
- [Optional] Is the output of the notebook unchanged?

There are 2 kinds of notebooks in the repository:
 - `examples` - containing executable Python code
   - some examples also contain Python mark-up snippets, can we ensure these do not go out-of-date?
 - documentation - shouldn't contain executable Python code, output upon execution should be unchanged
   - documentation may contain Python mark-up  snippets which ideally we would convert to executable code and test, however this is likely impossible as the snippets are used as illustration and not have any dependencies necessarily defined or loaded
  
Some options:
- A promising starting point is this simple pytest plugin: https://github.com/treebeardtech/nbmake
- An older plugin mostly for ensuring output consistency: https://github.com/chrisjsewell/pytest-notebook
- Another one in the same vein: https://github.com/computationalmodelling/nbval
- An alternative for more complex scenarios: https://papermill.readthedocs.io/en/latest/
- Roll own with `jupytext`/`nbconvert`

Since notebooks may take a long time to execute, we may create a separate GA workflow to be run e.g. once a week / before a release is due.

**Comment by jklaise** [comment]
Initial run on `examples` directory ran for ~1h 47min (serially) with 12/33 notebooks failing to finish:

- 9 timed out after a cell took 300 seconds to execute (review if we can reduce this time and/or extend default timeout)
- 1 `TypeError` in `integrated_gradients_imdb.ipynb` - genuine bug
- 1 `TypeError` in `cfproto_mnist.ipynb` - could be a genuine bug
- 1 `AttributeError` in `cem_mnist-search-for-good-examples.ipynb` - I think this may be a rogue development notebook that doesn't belong in the repo

**Comment by jklaise** [comment]
Some options to deal with notebooks that have cells timing out:
 - increase cell timeout (supported by `nbmake`), would need to check how long the longest-running notebook takes
 - skip notebook entirely (not natively supported by `nbmake` apart from listing every tested notebook manually, also not ideal to skip whole notebooks)
 - at test time parametrize cells/notebooks to use values that are smaller and result in shorter runs (could investigate `papermill` for this) - trickier than the other options, may not be possible in all cases

**Comment by alex-treebeard** [comment]
hi @jklaise -- I'm maintainer of nbmake.

nbmake, and other pytest plugins can use pytest's discovery flags to ignore test cases ([docs](https://docs.pytest.org/en/6.2.x/example/pythoncollection.html), [example](https://github.com/treebeardtech/nbmake/blob/51553313f2eea748ac1c349a8eb0c583032d9920/tests/test_pytest_plugin.py#L37))

Thanks for checking out the project, issues welcome.

**Comment by jklaise** [comment]
Tracking comment for notebooks that are timing out and proposed fixes (see this GA run https://github.com/SeldonIO/alibi/actions/runs/1254383581):
 - [x] 8 notebooks need `ipywidgets` which is bundled by `jupyter` but **not** by `jupyterlab` or in our dev dependencies. Fixed by https://github.com/SeldonIO/alibi/commit/fb028083b1a4c7d75ca7be6fe80e42567060e324
 -  [ ] `cem_mnist.ipynb.ipynb` - TBD
 -  [ ] `cfproto_mnist.ipynb` - TBD
 -  [ ] `trustscore_mnist.ipynb` - autoencoder epochs=8 is causing the timeout, could drop to 4 (in line with other MNIST notebooks), but need to check impact
 -  [ ] `xgboost_model_fitting_adult.ipynb` - extremely long-running hyperparameter tuning cells, best either to ignore this notebook explicitly or write it in a way that can be re-parametrized with `papermill`
 - [ ] `anchor_text_movie.ipynb` - auto-regressive language model explain time
 - [ ]  `cfrl_adult.ipynb.ipynb` - explainer fit time
 - [ ]  `cfrl_mnist.ipynb.ipynb` - autoencoder (and explainer) fit time
 - [ ] `distributed_kernel_shap_adult_lr.ipynb` - the cell running an explainer sequentially for 3 runs
 - [ ] `integrated_gradients_transformers.ipynb` - generating embeddings
 - [ ] `interventional_tree_shap_adult_xgb.ipynb` - cell running `KernelShap` to test convergence
 - [ ] `kernel_shap_adult_lr.ipynb` - cell calling `lr_explainer.explain`

**Comment by jklaise** [comment]
Unfortunately `papermill` does not support more than one parameter cell which could make it pretty awkward for parametrizing user-facing notebooks in multiple places: https://github.com/nteract/papermill/issues/328

**Comment by jklaise** [comment]
We've decided not to target 100% coverage as some notebooks are just more applied and involve steps that can't be easily (and quickly) replicated in a CI environment.

The next step to close this would be to adapt the current solution to exclude certain notebooks as well as extend maximum notebook cell runtime for those that can be run but just need a bit more time to finish.

**Comment by jklaise** [comment]
We want the testing framework to respect a list (that we maintain) of notebooks not to be executed. This list should be respected both on testing all notebooks as well as testing changed/new notebooks only on a PR.

This poses some difficulties as `nbmake` test discovery is command-line driven and it's difficult to exclude certain notebooks in both scenarios discussed above.

As a result, I think an approach using `pytest` would be best, although with a workaround required. The basic structure is in `testing/test_notebooks.py`:
 - Notebooks discovered using `glob`
 - Add a list to ignore certain notebooks
 - Notebook names passed as parameters in the test

The difficulty becomes in supporting the "test only changed notebooks" workflow. The usual way to achieve this is to specify test cases via the `-k` command line argument:

```bash
pytest testing/test_notebooks.py -k ''test_notebook_execution[examples/ale_classification.ipynb]"
```

However, this fails because `pytest` doesn't support `/` in test selection. A workaround would be to refactor the test so it only takes in the name of the notebook (i.e. `ale_classification.ipynb`) and add the path in front. To do this from the `test_changed_notebooks.yml` script we would need to use a shell one-liner to parse changed notebooks, e.g. `'examples/ale_classification.ipynb examples/ale_regression_boston.ipynb'` into `'ale_classification.ipynb or ale_regression_boston.ipynb'` (note the use of `or` which will discover notebooks with either name.

The main limitation of this is requiring all notebooks to be in the top level of `examples` which we are currently adhering to.

**Comment by jklaise** [comment]
Closed via #484.
